db-queue-spring/src/main/java/ru/yoomoney/tech/dbqueue/spring/dao/H2QueuePickTaskDao.java

```diff
@@ -0,0 +1,288 @@
+                            .withTotalAttemptsCount(rs.getLong(queueTableSchema.getTotalAttemptField()))
+                            .withExtData(additionalData).build();
+
+                }, location.getQueueId().asString(), pickTaskSettings.getRetryInterval().getSeconds());
+    }
+
+    public static class PickerProcedure {
+        private static final Map<String, PickQuery> pickTaskSqlCache = new ConcurrentHashMap<>();
+        private static final RowIdLocker rowIdLocker = new RowIdLocker();
+
+        public static ResultSet pickTask(final Connection connection,
+                                         final String queueName,
+                                         final long retryInterval) throws SQLException {
+            final PickQuery pickQuery = pickTaskSqlCache.get(queueName);
+            Objects.requireNonNull(pickQuery, "A pick query can't be null");
+
+            final Savepoint savepoint = connection.setSavepoint("start-picking-task-point");
+
+            final Long taskId = rowIdLocker.lock(queueName, rowIds -> {
+                try {
+
+                    final String sql;
+                    if (rowIds.isEmpty())
+                        sql = String.format(pickQuery.selectSql, " (1 = 1) ");
+                    else
+                        sql = String.format(
+                                pickQuery.selectSql,
+                                " _ROWID_ IN " + rowIds
+                                        .stream()
+                                        .map(Object::toString)
+                                        .collect(Collectors.joining(",", "(", ")")));
+
+                    PreparedStatement chooseStatement = connection.prepareStatement(sql);
+                    chooseStatement.setString(1, queueName);
+                    ResultSet chooseResultSet = chooseStatement.executeQuery();
+                    if (!chooseResultSet.next())
+                        return null;
+
+                    return chooseResultSet.getLong(1);
+                } catch (Exception e) {
+                    throw new IllegalStateException("can't select pick task", e);
+                }
+            });
+
+            if (taskId == null)
+                return null;
+
+            try {
+                PreparedStatement updateStatement = connection.prepareStatement(pickQuery.updateSql);
+                updateStatement.setLong(1, retryInterval);
+                updateStatement.setLong(2, taskId);
+                int updatedRowCount = updateStatement.executeUpdate();
+                if (updatedRowCount != 1)
+                    throw new IllegalStateException("Something wrong went here. Only row must be updated, not more!");
+
+                final PreparedStatement selectStatement = connection.prepareStatement(pickQuery.returnSql);
+                selectStatement.setLong(1, taskId);
+                final ResultSet resultSet = selectStatement.executeQuery();
+
+                connection.commit();
```

Comment:
I suppose this is an issue. Db-queue has several processing modes which manage transactions for example https://github.com/yoomoney-tech/db-queue/blob/master/db-queue-core/src/main/java/ru/yoomoney/tech/dbqueue/internal/runner/QueueRunnerInTransaction.java

This statement breaks the behavior of ProcessingMode#WRAP_IN_TRANSACTION

Reply:
I didn't know it. Thank you. I will bear it in mind